KEP-4222: Add CBOR variant of admission webhook integration test.

benluddy opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind cleanup
/sig api-machinery

#### What this PR does / why we need it:

The existing admission webhook integration test provides good coverage of serving built-in resources
and custom resources, including subresources. Serialization concerns, including roundtrippability,
of built-in types have existing test coverage; the CBOR variant of the admission webhook integration
test additionally exercises client and server codec wiring.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

As outlined in https://github.com/kubernetes/enhancements/tree/e1d4725d07a07c40da7ded6e1dd44c0f9aaeaeb9/keps/sig-api-machinery/4222-cbor-serializer#phased-implementation, the programmatic wiring to enable integration testing CBOR functionality is being added first, with runtime feature gating to follow only after all required tests are merged.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://kep.k8s.io/4222
```

benluddy committed:
Use runtime.SerializerInfo in place of internal "serializerType".

CodecFactory construction uses an unexported struct type named "serializerType" to hold serializer
definitions. There are few differences between it and runtime.SerializerInfo, and they do not appear
to be used anymore. For example, serializerType includes an unused FileExtensions field, and has
distinct ContentType (singular) and AcceptContentTypes (plural) fields instead of
runtime.SerializeInfo's singular MediaType. All remaining uses of serializerType set
AcceptContentTypes to a single-entry slice whose element is equal to its ContentType field.

During construction of a CodecFactory, all serializerType values were already being mechanically
translated into runtime.SerializerInfo values.

Moving to an exported type for serializer definitions makes it easier to expose an option to allow
callers to register their own serializer definitions, which in turn makes it possible to
conditionally include new serializers at runtime (especially behind feature gates).

benluddy committed:
Add WithSerializer option to add serializers to CodecFactory.

benluddy committed:
Add CBOR serializer option to disable JSON transcoding of raw types.

benluddy committed:
Export meta internal version scheme for testing.

Codecs is already exported, but in order for tests to construct an alternate CodecFactory for meta's
internal version types, they either need to be able to reference the scheme or to construct a
parallel scheme, and a parallel scheme construction risks going out of sync with the way the
package-scoped scheme object is initialized.

benluddy commented:
/triage accepted
/cc @jpbetz @deads2k

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go:
```diff
@@ -79,6 +80,15 @@ func Strict(s bool) Option {
+// Transcode configures a serializer to transcode the "raw" bytes of a decoded runtime.RawExtension
+// or metav1.FieldsV1 object to JSON. This is enabled by default to support existing programs that
+// depend on the assumption that objects of either type contain valid JSON.
+func Transcode(s bool) Option {
```
This godoc is great. Could we get similar godoc on `Strict` and `NewSerializer` so it's clear what all the defaults are?

jpbetz commented on staging/src/k8s.io/client-go/dynamic/scheme.go:
```diff
@@ -35,11 +35,10 @@ func init() {
 
-// basicNegotiatedSerializer is used to handle discovery and error handling serialization
-type basicNegotiatedSerializer struct{}
+var ExtraSerializers []func(runtime.ObjectCreater, runtime.ObjectTyper) runtime.SerializerInfo
```
Same as above. Global shared access to variables like this is something I'd like to avoid.

jpbetz commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -600,6 +600,8 @@ func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions
 	return info.storages[info.storageVersion].CustomResource, nil
 }
 
+var ExtraSerializers []func(runtime.ObjectCreater, runtime.ObjectTyper) runtime.SerializerInfo
```
Global shared access to variables like this is something I'd like to avoid.

jpbetz replied:
At the very minimum, could we funnel access through a `SetXyzForTests` type function that clearly documents the limitations?  Curious if there are better ways to do this that anyone knows of (@deads2k and @sttts for any ideas)

benluddy replied:
They should all be replaced by feature gate checks once the required integration tests are complete (https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4222-cbor-serializer#phased-implementation). What do you think about introducing a separate `FeatureGate` object for this (one that is not wired to flags at runtime)?

jpbetz commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -742,7 +742,7 @@ func (c *RecommendedConfig) Complete() CompletedConfig {
 }
 
-var allowedMediaTypes = []string{
+var AllowedMediaTypes = []string{
```
Global shared access to variables like this is something I'd like to avoid.

jpbetz replied:
For this case, is there some way we could make it possible to added the CBOR media type via configuration?

jpbetz commented on staging/src/k8s.io/apiserver/pkg/features/kube_features.go:
```diff
@@ -458,3 +464,12 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 // Please do not add to this struct and use defaultVersionedKubernetesFeatureGates instead.
 var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureSpec{}
+
+// testOnlyVersionedKubernetesFeatureGates consists of features that require programmatic enablement
+// for integration testing, but have not yet graduated to alpha in a release and must not be enabled
+// by a runtime option.
+var testOnlyVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate.VersionedSpecs{
+	CBORServingAndStorage: {
+		{Version: version.MustParse("1.32"), Default: false, PreRelease: featuregate.Alpha},
+	},
+}
```
I like the idea of leveraging feature gates. This feels more intuitive to me.

Building on this idea, what if we just put the feature gate directly into `defaultKubernetesFeatureGates` but named and documented it in a clear way, maybe `TestOnlyCBORServingAndStorage` with `PreRelease: featuregate.Deprecated` 
 with a comment explaining that this is temporary and only appropriate for use in test?

deads2k replied:
> Building on this idea, what if we just put the feature gate directly into `defaultKubernetesFeatureGates` but named and documented it in a clear way, maybe `TestOnlyCBORServingAndStorage` with `PreRelease: featuregate.Deprecated`
> with a comment explaining that this is temporary and only appropriate for use in test?

I'm hoping to prove the storage is correct and then place this into the normal gates.  The vast majority of features don't need this level of protection.  Are you ok with the approach knowing that basically only new storage encodings have ever needed this level of protection?

jpbetz replied:
> Are you ok with the approach knowing that basically only new storage encodings have ever needed this level of protection?

Yes, once this PR shifted away from exporting previously unexported variables and switched to feature gating I think I was satisfied.  The rest is just suggestions on how to communicate that the feature gates are "extra special", which, honestly, I don't feel super strong about... so long as it's reasonably clear to users what the nature of the feature gate is, I'm satisfied.

deads2k commented on test/integration/framework/cbor.go:
```diff
@@ -0,0 +1,152 @@
+		dynamic.ContentType = dynamicContentType
+		dynamic.ExtraSerializers = dynamicExtraSerializers
+	})
+	dynamic.AcceptContentTypes = "application/cbor;q=1,application/json;q=0.5"
```
what is the q?

benluddy replied:
Yes, without any quality parameters, it means all media types are as equally acceptable. We need to be sure that we don't depend on the way tiebreakers happen to be resolved.

> The weight is normalized to a real number in the range 0 through 1, where 0.001 is the least preferred and 1 is the most preferred; a value of 0 means "not acceptable". If no "q" parameter is present, the default weight is 1.

https://httpwg.org/specs/rfc9110.html#rfc.section.12.4.2

deads2k commented on test/integration/framework/cbor.go:
```diff
@@ -0,0 +1,152 @@
+				if len(requestbody) > 0 && !recognized {
+					t.Errorf("non-cbor request: 0x%x", requestbody)
+				}
+				request.Body = io.NopCloser(bytes.NewReader(requestbody))
```
please open a PR based on this one that `t.Error` here and in the cleanup below so it's obvious that these paths are being hit.

benluddy replied:
Opened https://github.com/kubernetes/kubernetes/pull/128295 and kicked off integration tests. I intentional fail to enable CBOR properly in that PR. Waiting for the presubmit output, but locally I see a lot of this:

```
cbor.go:110: non-cbor request: 0x7b226b696e64223a2244656c6574654f7074696f6e73222c2261706956657273696f6e223a227631222c226772616365506572696f645365636f6e6473223a302c2270726f7061676174696f6e506f6c696379223a224261636b67726f756e64227d0a
cbor.go:136: non-cbor response: 0x7b226b696e64223a22537461747573222c2261706956657273696f6e223a227631222c226d65746164617461223a7b7d2c22737461747573223a224661696c757265222c226d657373616765223a22766f6c756d6561747472696275746573636c61737365732e73746f726167652e6b38732e696f205c22766163325c22206e6f7420666f756e64222c22726561736f6e223a224e6f74466f756e64222c2264657461696c73223a7b226e616d65223a2276616332222c2267726f7570223a2273746f726167652e6b38732e696f222c226b696e64223a22766f6c756d6561747472696275746573636c6173736573227d2c22636f6465223a3430347d0a
```

They're both JSON, the first is:
```json
{"kind":"DeleteOptions","apiVersion":"v1","gracePeriodSeconds":0,"propagationPolicy":"Background"}
```
and the second is:
```json
{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"volumeattributesclasses.storage.k8s.io \"vac2\" not found","reason":"NotFound","details":{"name":"vac2","group":"storage.k8s.io","kind":"volumeattributesclasses"},"code":404}
```

deads2k replied:
great.

benluddy commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/codec_factory.go:
```diff
@@ -184,46 +169,26 @@ func NewCodecFactory(scheme *runtime.Scheme, mutators ...CodecFactoryOptionsMuta
+		}
+		alreadyAccepted[d.MediaType] = struct{}{}
+
+		if d.StreamSerializer != nil {
+			cloned := *d.StreamSerializer
+			d.StreamSerializer = &cloned
+		}
+		accepts = append(accepts, d)
```
I'm not sure the StreamSerializer copy in the original is anything more than a consequence of hoisting `serializerType`'s `StreamSerializer`, `EncodesAsText`, and `Framer` into a `StreamSerializerInfo`. I clone the input `*StreamSerializerInfo` since anyone who passed a `WithSerializer` option might have saved a reference to it. I doubt it would ever be a problem in practice and I would not feel bad removing the copy here.

Loop variables are scoped per-iteration since Go 1.22, so it shouldn't be necessary to also copy `d` (it still feels wrong to me because I am so used to the previous Go behavior).

deads2k replied:
>so it shouldn't be necessary to also copy d (it still feels wrong to me because I am so used to the previous Go behavior).

Ok, let's do it anyway for readability.

benluddy committed:
Wire test-only feature gate for CBOR serving.

To mitigate the risk of introducing a new protocol, integration tests for CBOR will be written using
a test-only feature gate instance that is not wired to runtime options. On alpha graduation, the
test-only feature gate instance will be replaced by a normal feature gate in the existing apiserver
feature gate instance.

benluddy committed:
Add test-only client feature gates for CBOR.

As with the apiserver feature gate for CBOR as a serving and storage encoding, the client feature
gates for CBOR are being initially added through a test-only feature gate instance that is not wired
to environment variables or to command-line flags and is intended only to be enabled
programmatically from integration tests. The test-only instance will be removed as part of alpha
graduation and replaced by conventional client feature gating.

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -755,6 +755,10 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 	if c.Serializer == nil {
 		return nil, fmt.Errorf("Genericapiserver.New() called with config.Serializer == nil")
 	}
+	allowedMediaTypes := allowedMediaTypes // shadow
```
find a different construct.  I can't tell what this is trying to do.

deads2k commented on staging/src/k8s.io/client-go/features/features.go:
```diff
@@ -141,3 +143,37 @@ var (
 	featureGates = &atomic.Value{}
 )
+
+// TODO: explain
```
yeah

benluddy committed:
Add CBOR variant of admission webhook integration test.

The existing admission webhook integration test provides good coverage of serving built-in resources
and custom resources, including subresources. Serialization concerns, including roundtrippability,
of built-in types have existing test coverage; the CBOR variant of the admission webhook integration
test additionally exercises client and server codec wiring.

benluddy commented:
/retest

benluddy commented:
/retest

deads2k commented:
The testfeaturegate approach avoids expanding globals.  The test PR demonstrates this works.  Great work moving this forward.  I think we still need a test reading the etcd content to confirm CBOR.

/lgtm
/approve
